source/agora/test/GossipProtocol.d

```diff
@@ -37,33 +40,54 @@ unittest
         assert(block_height == 0, block_height.to!string);
     }
 
-    Transaction tx =
+    // Creates the first transaction.
+    Transaction firstTx = newCoinbaseTX(key_pairs[0].address, 100);
+    Hash firstHash = hashFull(firstTx);
+
+    node_1.putTransaction(firstTx);
+    // check block height
+    foreach (key, ref node; network.apis)
     {
-        [Input(hashFull("Message No. 1"))],
-        [Output(100)]
-    };
+        auto block_height = node.getBlockHeight();
+        assert(block_height == 1, block_height.to!string);
+    }
 
-    Hash tx_hash = hashFull(tx);
-    auto node_1 = network.apis.values[0];
-    node_1.putTransaction(tx);
+    // Creates the second transaction.
+    Transaction secondTx = Transaction(
+        [
+            Input(firstHash, 0)
+        ],
+        [
+            Output(50, key_pairs[1].address)
+        ]
+    );
+    Hash secondHash = hashFull(secondTx);
 
+    // It is validated. (the sum of `Output` <= the sum of `Input`)
+    node_1.putTransaction(secondTx);
     // Check hasTransactionHash
     foreach (key, ref node; network.apis)
     {
-        assert(node.hasTransactionHash(tx_hash));
+        //  every nodes received tx
+        assert(node.hasTransactionHash(secondHash));
 
         auto block_height = node.getBlockHeight();
-        assert(block_height == 1, block_height.to!string);
+        assert(block_height == 2, block_height.to!string);
     }
 
-    tx.inputs[0].previous = hashFull("Message No. 2");
-    tx_hash = hashFull(tx);
-    node_1.putTransaction(tx);
-
-    // check if the node height changed
+    // It isn't validated. (the sum of `Output` > the sum of `Input`)
+    secondTx.outputs ~= Output(100, key_pairs[2].address);
+    secondHash = hashFull(secondTx);
+    node_1.putTransaction(secondTx);
+    // Check hasTransactionHash
     foreach (key, ref node; network.apis)
     {
-        assert(node.hasTransactionHash(tx_hash));
+        //  first node received tx, others didn't receive tx
+        if (node == node_1)
+            assert(node.hasTransactionHash(secondHash));
```

Comment:
If a transaction is invalid, it should be discarded

Reply:
Modified